presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java

```diff
@@ -232,26 +228,28 @@ public RelationPlan planExpand(Query query)
 
         PlanNode replacementSpot = anchorPlan.getRoot();
         PlanNode recursionStep = coercedRecursionStep.getNode();
+        List<Symbol> mappings = coercedRecursionStep.getFields();
 
         // unroll recursion
         int maxRecursionDepth = getMaxRecursionDepth(session);
         for (int i = 0; i < maxRecursionDepth; i++) {
-            recursionSteps.add(copy(recursionStep));
-            PlanNode replacement = copy(recursionStep);
+            recursionSteps.add(alignOutputs(copy(recursionStep, mappings)));
```

Comment:
I changed it so that mappings are passed along with the PlanNode and whenever we need to bind symbols properly (that is, when replacing and when building UnionNode), it is done via mapping.